include/dice/sparse_map/internal/sparse_hash.hpp

```diff
@@ -0,0 +1,943 @@
+			pointer operator->() const noexcept { return &KeyValueSelect::both(*bucket_it_); }
+
+			sparse_iterator &operator++() noexcept {
+				assert(bucket_it_ != nullptr);
+				++bucket_it_;
+
+				if (bucket_it_ != (*cur_bucket_).end()) {
+					return *this;
+				}
+
+				do {
+					if (++cur_bucket_ == end_bucket_) {
+						bucket_it_ = nullptr;
+						return *this;
+					}
+				} while ((*cur_bucket_).empty());
+
+				bucket_it_ = (*cur_bucket_).begin();
+				return *this;
+			}
+
+			sparse_iterator operator++(int) noexcept {
+				auto tmp = *this;
+				++*this;
+				return tmp;
+			}
+
+			template<bool OIsConst>
+			bool operator==(sparse_iterator<OIsConst> const &other) const noexcept {
+				return cur_bucket_ == other.cur_bucket_ && bucket_it_ == other.bucket_it_;
+			}
+
+			template<bool OIsConst>
+			bool operator!=(sparse_iterator<OIsConst> const &other) const noexcept {
+				return cur_bucket_ != other.cur_bucket_ || bucket_it_ != other.bucket_it_;
+			}
+
+			template<bool OIsConst>
+			bool operator<(sparse_iterator<OIsConst> const &other) const noexcept {
+				return cur_bucket_ < other.cur_bucket_ || (cur_bucket_ == other.cur_bucket_ && bucket_it_ < other.bucket_it_);
+			}
+		};
+
+		iterator mutable_iterator(const_iterator pos) noexcept {
+			// SAFETY: this is non-const therefore the underlying buckets are also non-const
+			// as evidenced by the fact that we can call begin on them
+			auto it_sparse_buckets = buckets_.begin() + std::distance(buckets_.cbegin(), pos.cur_bucket_);
+
+			// SAFETY: this is non-const therefore the underlying sparse array is also non-const
+			auto it_array = sparse_bucket_type::unsafe_mutable_iterator(pos.bucket_it_);
+
+			return iterator{it_sparse_buckets, buckets_.end(), it_array};
+		}
+
+	public:
+		sparse_hash(size_type bucket_count,
+					hasher const &hash,
+					key_equal const &equal,
+					allocator_type const &alloc) : h_{hash},
+												   keq_{equal},
+												   gpol_{bucket_count},
+												   buckets_{bucket_count, alloc},
+												   bucket_count_{bucket_count},
+												   n_elements_{0},
+												   n_deleted_elements_{0},
+												   load_threshold_rehash_{calc_load_threshold_rehash(bucket_count)},
+												   load_threshold_clear_deleted_{calc_load_threshold_clear_deleted(bucket_count)} {
+
+			// Check in the constructor instead of outside of a function to avoid
+			// compilation issues when value_type is not complete.
+			static_assert(std::is_nothrow_move_constructible<value_type>::value ||
+								  std::is_copy_constructible<value_type>::value,
+						  "Key, and T if present, must be nothrow move constructible "
+						  "and/or copy constructible.");
+		}
+
+		sparse_hash(sparse_hash const &other) = default;
+		sparse_hash &operator=(sparse_hash const &other) = default;
+
+		sparse_hash(sparse_hash &&other) noexcept(std::is_nothrow_move_constructible_v<sparse_bucket_array_type>
```

Comment:
I think this should be the case in general if `std::is_nothrow_move_constructible<value_type>::value` (and that is guaranteed in the constructor above).

Reply:
Even better, this is always true actually